client-src/index.js

```diff
@@ -266,17 +301,24 @@ const onSocketMessage = {
       log.warn(printableWarnings[i]);
     }
 
-    const needShowOverlayForWarnings =
+    const overlayWarningsSetting =
       typeof options.overlay === "boolean"
         ? options.overlay
         : options.overlay && options.overlay.warnings;
 
-    if (needShowOverlayForWarnings) {
-      overlay.send({
-        type: "BUILD_ERROR",
-        level: "warning",
-        messages: warnings,
-      });
+    if (overlayWarningsSetting) {
+      const warningsToDisplay =
+        typeof overlayWarningsSetting === "function"
+          ? warnings.filter(overlayWarningsSetting)
+          : warnings;
+
+      if (warningsToDisplay.length) {
+        overlay.send({
+          type: "BUILD_ERROR",
+          level: "warning",
+          messages: warnings,
```

Comment:
Shouldn't this be `BUILD_WARNING`?

Reply:
the event type purpose is to determine how to transition the state:

https://github.com/webpack/webpack-dev-server/blob/f7386f60da170b158d13a2c41cc668f0b83d9679/client-src/overlay/state-machine.js#L32-L38

I don't mind refactor to add event type `BUILD_WARNING` and remove `level` property, although it does introduce some duplication in the state machine.

Reply:
I think it's fine 👍🏻